### PR TITLE
fix(frontend/controller): improve match highlighting + split up the search component

### LINKED ISSUE #567: Broken search results highlighter for matching hindi/urdu translits
**Screenshots**
image

image

image

**Additional context**
Inspecting does show that some "matched" span is added, without any text.

### LINKED ISSUE #569: Last word repeating in search results
**Describe the bug**
When transliterations are turned on for search results. Last word is repeated twice.

**To Reproduce**
1. Go to 'Settings'
2. Click on 'Activities' > 'Search'
3. Select 'English' in 'Transliterations'
4. Go back to controller and search 'qq'
5. See the last word is repeated for everything

**Screenshots**
image

**Specs**
 - Version: Dev branch head at 0a990ac

### PR DESCRIPTION
**Pull request by saihaj**
### Summary of PR
The Results function was very huge and it was getting quite confusing to navigate around. Idea here is to split up the search component so that results becomes its own component and all highlighting is contained there. Search component will pass down all the necessary props. 

I tried bringing context down to results (where it makes way much more sense and then we don't have to pass so many things down) but was not possible because it breaks react hooks rules. Later I realized it is also not so great because we are mapping though the list of results we get from backend and that might be too much work each time to get context from the hook. 

@Harjot1Singh you probably have some better way because this has reached very high levels of prop drilling.

* [x] Reolve => [highlighting of a search results when keyboard keys are pressed to select an item](https://github.com/ShabadOS/desktop/pull/594#issuecomment-660258042)

### Fixes
Fixes #567 
Fixes #569 

### TTR

@saihaj - unsure
@Harjot1Singh - 2 hours

### CODE REVIEW

**Comment by saihaj** [comment]
@Harjot1Singh I need help, I messed up highlighting of a search results when keyboard keys are pressed to select an item.

**Comment by Harjot1Singh** [comment]
Please check this still works well @bhajneet @saihaj.

I ended up fixing some search highlighting issues, and refactoring the logic into a separate file.

**Comment by saihaj** [comment]
✅ LGTM
This also fixes #567 #569

File: app/frontend/src/Controller/Search/Result.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * Renders a single result, highlighting the match.
+ * @param {string} gurmukhi The shabad line to display.
+ * @param {int} typeId The type id of line.
+ * @param {string} lineId The id of the line.
+ * @param {string} shabadId The id of the shabad.
+ * @param {Component} ref The ref to the component.
+ * @param {int} sourceId The id of source.
+ * @param {Object} shabad The object containng section information and other metadata.
```
Comment (saihaj) [inline app/frontend/src/Controller/Search/Result.js:19]:
```suggestion
 * @param {Object} shabad The object containing section information and other metadata.
```

File: app/frontend/src/Controller/Search/match-highlighter.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  const baseGurmukhi = stripVishraams( gurmukhi )
+  const baseTarget = stripVishraams( target )
+
+  const sanitisedQuery = query.trim()
+
+  const foundPosition = baseGurmukhi.search( sanitisedQuery )
+  const matchStartPosition = baseGurmukhi.lastIndexOf( ' ', foundPosition )
+
+  const wordEndPosition = baseGurmukhi.indexOf( ' ', foundPosition + sanitisedQuery.length )
+  // If the match finishes in the last word, no space will be deteced, and wordEndPosition
```
Comment (saihaj) [inline app/frontend/src/Controller/Search/match-highlighter.js:15]:
```suggestion
  // If the match finishes in the last word, no space will be detected, and wordEndPosition
```

File: app/frontend/src/Controller/Search/match-highlighter.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * @return An array of [ beforeMatch, match, afterMatch ],
+ *   with `match` being the highlighted section.`.
+ */
+const getHighlighter = ( searchQuery, searchMode ) => context => target => {
+  if ( !target ) return [ '', '', '' ]
+
+  // Account for wildcard characters
+  const sanitizedQuery = searchQuery.replace( new RegExp( '_', 'g' ), '.' )
+
+  // Select the right higlighter
```
Comment (saihaj) [inline app/frontend/src/Controller/Search/match-highlighter.js:71]:
```suggestion
  // Select the right highlighter
```

**Comment by Harjot1Singh** [comment]
This serves as a very good basis and the preliminary work required by #221, which will serve as one of our biggest performance improvements.

File: app/frontend/src/Controller/Search/Result.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+/**
+ * Renders a single result, highlighting the match.
+ * @param {string} gurmukhi The shabad line to display.
+ * @param {int} typeId The type id of line.
+ * @param {string} lineId The id of the line.
+ * @param {string} shabadId The id of the shabad.
+ * @param {Component} ref The ref to the component.
+ * @param {int} sourceId The id of source.
+ * @param {Object} shabad The object containing section information and other metadata.
```
Comment (bhajneet) [inline app/frontend/src/Controller/Search/Result.js:19]: Doesn't match up below

**Review by bhajneet — DISMISSED** [review]
Great work! Would have liked a little more narrative with comments. Working fine with tests of wildcard characters, full words, and all options in search (translit/translation/citation, etc.). LGTM!

**Review by Harjot1Singh — APPROVED** [review]

**Review by Harjot1Singh — APPROVED** [review]

**Pull request merged**